sycl/doc/extensions/proposed/sycl_ext_oneapi_composite_device.asciidoc

```diff
@@ -0,0 +1,290 @@
+tile it runs on.
+
+Applications that use this extension can access the same hardware through
+different `device` objects.  For example, it is possible to get one `device`
+object that represents a card and another `device` object that represents a
+tile from that card.  As noted previously, a device representing a card is
+known as a "composite" device.  A device representing a tile is known as a
+"component" of that composite device.  Since the tile-level devices are
+returned from `device::get_devices`, they are also considered "root" devices,
+which is a term defined in the core SYCL specification.
+
+
+== Specification
+
+=== Feature test macro
+
+This extension provides a feature-test macro as described in the core SYCL
+specification.  An implementation supporting this extension must predefine the
+macro `SYCL_EXT_ONEAPI_COMPOSITE_DEVICE` to one of the values defined in the
+table below.  Applications can test for the existence of this macro to
+determine if the implementation supports this feature, or applications can test
+the macro's value to determine which of the extension's features the
+implementation supports.
+
+[%header,cols="1,5"]
+|===
+|Value
+|Description
+
+|1
+|The APIs of this experimental extension are not versioned, so the
+ feature-test macro always has this value.
+|===
+
+=== Enumerating composite devices
+
+This extension adds two new functions for enumerating the available composite
+devices.  One is a free function and the other is a new member function of the
+`platform` class:
+
+```
+namespace sycl {
+namespace ext::oneapi::experimental {
+
+std::vector<device> get_composite_devices();
+
+} // namespace ext::oneapi::experimental
+
+class platform {
+  std::vector<device> ext_oneapi_get_composite_devices();
+};
+
+} // namespace sycl
+```
+
+The free function `get_composite_devices` returns all of the composite devices
+across all platforms.  The member function
+`platform::ext_oneapi_get_composite_devices` returns the composite devices
+within the given platform.
+
+The execution environment for a SYCL application has a fixed number of
+composite devices which does not vary as the application executes.  As a
+result, each call to these functions returns the same set of `device` objects,
+and the order of those objects does not vary between calls.
+
+=== New device information descriptors
+
+This extension adds two new device information descriptors:
+
+```
+namespace sycl::ext::oneapi::experimental::info::device {
+
+struct component_devices;
+struct composite_device;
+
+} // namespace sycl::ext::oneapi::experimental::info::device
+```
+
+[width="100%",%header,cols="37%,19%,44%"]
+|===
+|Device descriptor
+|Return type
+|Description
+
+|`component_devices`
+|`std::vector<device>`
+|Returns the set of component devices that are contained by a composite device.
+ If "this" device is not a composite device, an empty vector is returned.
```

Comment:
Why is it "this" device?  This looks odd to me, and I don't think we use this style anywhere else.

Reply:
I use this style when I think there could be some confusion about which device I'm talking about.

There's probably more potential confusion with the `composite_device` query because that query also returns a device.  By saying "this" device, I'm trying to emphasize that I'm talking about the device that is the C++ "this" object, and not the device that's returned by the query.

If that meaning did not come across, I can try to find some alternate wording.  FWIW, I've occasionally used this wording in the core SYCL spec also.

Reply:
Personally, I find it more confusing than helpful, because there are more instances of the SYCL specification where we _don't_ follow this convention.  Several of the entries in Table 23 say "this `device`", and I've always thought that clearly meant "the `device` object the member function was called from".

I think my preference here would be to lean the other way, and avoid saying "this `device`" if we're referring to something else.  If a function takes a `device` as an argument, we can unambiguously refer to it by name.  If a function returns a `device`, we can unambiguously refer to it as "the return value".

I'm not opposed to adopting a different convention through the SYCL specification and our extensions (which could be "this device", or `this`, or something else) but I think mixing conventions is confusing.  We should probably also look at how ISO C++ solves this problem for member functions.

Reply:
OK, I removed use of `"this"` in 1f1712ed025c76a30ca251c97175685a59bf58bb.

I also checked for other places I used this convention.  The only other place is in the core SYCL spec in KhronosGroup/SYCL-Docs#514.  I removed the `"this"` from that PR also.  See KhronosGroup/SYCL-Docs@cd14ca9417196406c04eb7cb15a50f5105d30f16

Reply:
Thanks!

I took a quick look through ISO C++, and I can see some instances of "this `classname` object" and some of `*this`.  I haven't quite figured out the convention: I think it might be that they use the former when referring to a property of am object (e.g., "The stored exception captured by this `nested_exception` object") and the latter when referring to an entire abstract object (e.g., "Returns `true` if `*this` precedes `rhs` in the implementation's collation order.")